File: 06 - LOGICIEL EMBARQUE/simulateur/src/history-log.js
export const HISTORY_RECORD_SIZE = 24;
export const EVENT_RECORD_SIZE = 32;

export const HISTORY_STATE = Object.freeze({
  STARTUP: 0, AUTO: 1, TEST: 2, FORCE_OPEN: 3, FAULT: 4,
  DEGRADED_FLOW: 5, UPDATE: 6, RECOVERY: 7,
});

export const HISTORY_FAULT = Object.freeze({
  NONE: 0, SENSOR_MISSING: 1, SENSOR_RANGE: 2, SENSOR_STALE: 3,
  HIGH_CO2: 4, OUTPUT_WRITE_FAILED: 5, CONFIG_CORRUPT: 6,
  WIFI_START_FAILED: 7, HISTORY_STORAGE_FAILED: 8,
});

export const HISTORY_FLAG = Object.freeze({
  TIME_SYNCED: 1, SENSOR_VALID: 2, FLOW_VALID: 4, TEST_ACTIVE: 8,
});

export const EVENT_CODE = Object.freeze({
  BOOT: 1, CONFIG_CHANGED: 2, CONFIG_RECOVERY: 3, SERVICE_OPENED: 4,
  SERVICE_CLOSED: 5, LOGIN_FAILED: 6, TEST_STARTED: 7, TEST_STOPPED: 8,
  FAULT_RAISED: 9, FAULT_CLEARED: 10, FIRMWARE_UPDATE: 11,
  HISTORY_STORAGE_FAILED: 12, SESSION_OPENED: 13, SESSION_CLOSED: 14,
  HISTORY_CLEARED: 15,
});

export const EVENT_SEVERITY = Object.freeze({ INFO: 0, WARNING: 1, ERROR: 2, CRITICAL: 3 });
export const EVENT_SOURCE = Object.freeze({ SYSTEM: 0, PHYSICAL: 1, INSTALLER: 2, ADMIN: 3, GTB: 4 });

const reverse = (object) => Object.fromEntries(Object.entries(object).map(([key, value]) => [value, key]));
const STATE_NAME = reverse(HISTORY_STATE);
const FAULT_NAME = reverse(HISTORY_FAULT);
const EVENT_NAME = reverse(EVENT_CODE);
const SEVERITY_NAME = reverse(EVENT_SEVERITY);
const SOURCE_NAME = reverse(EVENT_SOURCE);

export function crc16Ccitt(bytes) {
  let crc = 0xffff;
  for (const byte of bytes) {
    crc ^= byte << 8;
    for (let bit = 0; bit < 8; bit += 1) crc = (crc & 0x8000) ? ((crc << 1) ^ 0x1021) : (crc << 1);
    crc &= 0xffff;
  }
  return crc;
}

function enumValue(map, value, field) {
  if (typeof value === "number" && Object.values(map).includes(value)) return value;
  if (typeof value === "string" && Object.hasOwn(map, value)) return map[value];
  throw new RangeError(`${field} inconnu.`);
}

function uint(value, max, field) {
  if (!Number.isInteger(value) || value < 0 || value > max) throw new RangeError(`${field} invalide.`);
  return value;
}

function int32(value, field) {
  if (!Number.isInteger(value) || value < -0x80000000 || value > 0x7fffffff) throw new RangeError(`${field} invalide.`);
  return value;
}

export function encodeHistoryRecord(sample) {
  const bytes = new Uint8Array(HISTORY_RECORD_SIZE);
  const view = new DataView(bytes.buffer);
  const co2 = sample.co2Ppm == null ? 0xffff : uint(sample.co2Ppm, 10000, "co2Ppm");
  const outputTenths = Math.round(Number(sample.outputPct) * 10);
  if (!Number.isFinite(sample.outputPct) || Math.abs(sample.outputPct * 10 - outputTenths) > 1e-9) throw new RangeError("outputPct invalide.");
  view.setUint32(0, uint(sample.sequence, 0xffffffff, "sequence"), true);
  view.setUint32(4, uint(sample.epochSeconds ?? 0, 0xffffffff, "epochSeconds"), true);
  view.setUint32(8, uint(sample.uptimeSeconds, 0xffffffff, "uptimeSeconds"), true);
  view.setUint16(12, uint(sample.bootId, 0xffff, "bootId"), true);
  view.setUint16(14, co2, true);
  view.setUint16(16, uint(outputTenths, 1000, "outputPct"), true);
  view.setUint8(18, enumValue(HISTORY_STATE, sample.state, "state"));
  view.setUint8(19, enumValue(HISTORY_FAULT, sample.fault, "fault"));
  view.setUint16(20, uint(sample.flags ?? 0, 0xffff, "flags"), true);
  view.setUint16(22, crc16Ccitt(bytes.subarray(0, 22)), true);
  return bytes;
}

export function decodeHistoryRecord(input) {
  const bytes = input instanceof Uint8Array ? input : new Uint8Array(input ?? []);
  if (bytes.length !== HISTORY_RECORD_SIZE) return { ok: false, reason: "LENGTH" };
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  if (view.getUint16(22, true) !== crc16Ccitt(bytes.subarray(0, 22))) return { ok: false, reason: "CRC" };
  const state = STATE_NAME[view.getUint8(18)];
  const fault = FAULT_NAME[view.getUint8(19)];
  if (!state || !fault) return { ok: false, reason: "ENUM" };
  const co2 = view.getUint16(14, true);
  return {
    ok: true,
    sample: {
      sequence: view.getUint32(0, true), epochSeconds: view.getUint32(4, true),
      uptimeSeconds: view.getUint32(8, true), bootId: view.getUint16(12, true),
      co2Ppm: co2 === 0xffff ? null : co2, outputPct: view.getUint16(16, true) / 10,
      state, fault, flags: view.getUint16(20, true),
    },
  };
}

export function encodeEventRecord(event) {
  const bytes = new Uint8Array(EVENT_RECORD_SIZE);
  const view = new DataView(bytes.buffer);
  view.setUint32(0, uint(event.sequence, 0xffffffff, "sequence"), true);
  view.setUint32(4, uint(event.epochSeconds ?? 0, 0xffffffff, "epochSeconds"), true);
  view.setUint32(8, uint(event.uptimeSeconds, 0xffffffff, "uptimeSeconds"), true);
  view.setUint16(12, uint(event.bootId, 0xffff, "bootId"), true);
  view.setUint16(14, enumValue(EVENT_CODE, event.code, "code"), true);
  view.setUint8(16, enumValue(EVENT_SEVERITY, event.severity, "severity"));
  view.setUint8(17, enumValue(EVENT_SOURCE, event.source, "source"));
  view.setUint16(18, uint(event.count ?? 1, 0xffff, "count"), true);
  view.setInt32(20, int32(event.detail0 ?? 0, "detail0"), true);
  view.setInt32(24, int32(event.detail1 ?? 0, "detail1"), true);
  view.setUint16(28, 0, true);
  view.setUint16(30, crc16Ccitt(bytes.subarray(0, 30)), true);
  return bytes;
}

export function decodeEventRecord(input) {
  const bytes = input instanceof Uint8Array ? input : new Uint8Array(input ?? []);
  if (bytes.length !== EVENT_RECORD_SIZE) return { ok: false, reason: "LENGTH" };
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  if (view.getUint16(28, true) !== 0) return { ok: false, reason: "RESERVED" };
  if (view.getUint16(30, true) !== crc16Ccitt(bytes.subarray(0, 30))) return { ok: false, reason: "CRC" };
  const code = EVENT_NAME[view.getUint16(14, true)];
  const severity = SEVERITY_NAME[view.getUint8(16)];
  const source = SOURCE_NAME[view.getUint8(17)];
  if (!code || !severity || !source) return { ok: false, reason: "ENUM" };
  return {
    ok: true,
    event: {
      sequence: view.getUint32(0, true), epochSeconds: view.getUint32(4, true),
      uptimeSeconds: view.getUint32(8, true), bootId: view.getUint16(12, true),
      code, severity, source, count: view.getUint16(18, true),
      detail0: view.getInt32(20, true), detail1: view.getInt32(24, true),
    },
  };
}

class CircularLog {
  constructor(capacity) {
    this.capacity = uint(capacity, 1_000_000, "capacity");
    if (capacity < 1) throw new RangeError("capacity invalide.");
    this.records = [];
  }
  append(record) {
    if (this.records.length === this.capacity) this.records.shift();
    this.records.push(record);
  }
  latest(limit = this.capacity) { return this.records.slice(-Math.max(0, Math.min(this.capacity, limit))); }
  get size() { return this.records.length; }
}

export class HistoryLog extends CircularLog {
  constructor({ samplePeriodSeconds = 60, retentionDays = 7, capacity } = {}) {
    const computedCapacity = Math.ceil(retentionDays * 86400 / samplePeriodSeconds);
    super(capacity ?? computedCapacity);
    this.samplePeriodSeconds = samplePeriodSeconds;
    this.nextSequence = 0;
    this.lastByBoot = new Map();
  }
  add(sample) {
    const previousUptime = this.lastByBoot.get(sample.bootId);
    if (previousUptime !== undefined && sample.uptimeSeconds < previousUptime) throw new RangeError("uptime décroissant pour le même bootId.");
    const normalized = { ...sample, sequence: this.nextSequence >>> 0, epochSeconds: sample.epochSeconds ?? 0 };
    const decoded = decodeHistoryRecord(encodeHistoryRecord(normalized));
    if (!decoded.ok) throw new Error(decoded.reason);
    this.append(decoded.sample);
    this.lastByBoot.set(sample.bootId, sample.uptimeSeconds);
    this.nextSequence = (this.nextSequence + 1) >>> 0;
    return decoded.sample;
  }
}

export class EventLog extends CircularLog {
  constructor({ capacity = 512, coalesceWindowSeconds = 60 } = {}) {
    super(capacity);
    this.coalesceWindowSeconds = coalesceWindowSeconds;
    this.nextSequence = 0;
  }
  add(event) {
    const allowedKeys = new Set(["sequence", "epochSeconds", "uptimeSeconds", "bootId", "code", "severity", "source", "count", "detail0", "detail1"]);
    if (!event || typeof event !== "object" || Object.keys(event).some((key) => !allowedKeys.has(key))) {
      throw new TypeError("Les textes libres, secrets, identifiants et champs inconnus ne sont pas admis dans le journal persistant.");
    }
    const normalized = { ...event, sequence: this.nextSequence >>> 0, epochSeconds: event.epochSeconds ?? 0, count: 1 };
    const previous = this.records.at(-1);
    const canCoalesce = previous && normalized.severity !== "CRITICAL"
      && previous.bootId === normalized.bootId && previous.code === normalized.code
      && previous.severity === normalized.severity && previous.source === normalized.source
      && previous.detail0 === (normalized.detail0 ?? 0) && previous.detail1 === (normalized.detail1 ?? 0)
      && normalized.uptimeSeconds >= previous.uptimeSeconds
      && normalized.uptimeSeconds - previous.uptimeSeconds <= this.coalesceWindowSeconds;
    if (canCoalesce) {
      previous.count = Math.min(0xffff, previous.count + 1);
      previous.uptimeSeconds = normalized.uptimeSeconds;
      previous.epochSeconds = normalized.epochSeconds;
      return previous;
    }
    const decoded = decodeEventRecord(encodeEventRecord(normalized));
    if (!decoded.ok) throw new Error(decoded.reason);
    this.append(decoded.event);
    this.nextSequence = (this.nextSequence + 1) >>> 0;
    return decoded.event;
  }
}

export function estimateLogStorage({ samplePeriodSeconds = 60, retentionDays = 7, eventCapacity = 512 } = {}) {
  const sampleCount = Math.ceil(retentionDays * 86400 / samplePeriodSeconds);
  const historyBytes = sampleCount * HISTORY_RECORD_SIZE;
  const eventBytes = eventCapacity * EVENT_RECORD_SIZE;
  return { sampleCount, historyBytes, eventBytes, totalRecordBytes: historyBytes + eventBytes };
}

File: 06 - LOGICIEL EMBARQUE/firmware-esp32/components/history_log/history_log.cpp
#include "history_log.hpp"

#include <cmath>

namespace ventilation {
namespace {

void write_u16_le(uint8_t* data, size_t offset, uint16_t value) {
  data[offset] = static_cast<uint8_t>(value & 0xffU);
  data[offset + 1U] = static_cast<uint8_t>((value >> 8U) & 0xffU);
}

void write_u32_le(uint8_t* data, size_t offset, uint32_t value) {
  data[offset] = static_cast<uint8_t>(value & 0xffU);
  data[offset + 1U] = static_cast<uint8_t>((value >> 8U) & 0xffU);
  data[offset + 2U] = static_cast<uint8_t>((value >> 16U) & 0xffU);
  data[offset + 3U] = static_cast<uint8_t>((value >> 24U) & 0xffU);
}

void write_i32_le(uint8_t* data, size_t offset, int32_t value) {
  write_u32_le(data, offset, static_cast<uint32_t>(value));
}

uint16_t read_u16_le(const uint8_t* data, size_t offset) {
  return static_cast<uint16_t>(data[offset]) |
         static_cast<uint16_t>(static_cast<uint16_t>(data[offset + 1U]) << 8U);
}

uint32_t read_u32_le(const uint8_t* data, size_t offset) {
  return static_cast<uint32_t>(data[offset]) |
         (static_cast<uint32_t>(data[offset + 1U]) << 8U) |
         (static_cast<uint32_t>(data[offset + 2U]) << 16U) |
         (static_cast<uint32_t>(data[offset + 3U]) << 24U);
}

int32_t read_i32_le(const uint8_t* data, size_t offset) {
  return static_cast<int32_t>(read_u32_le(data, offset));
}

bool valid_state(uint8_t value) {
  return value <= static_cast<uint8_t>(HistoryState::kRecovery);
}

bool valid_fault(uint8_t value) {
  return value <= static_cast<uint8_t>(HistoryFault::kHistoryStorageFailed);
}

bool valid_event_code(uint16_t value) {
  return value >= static_cast<uint16_t>(EventCode::kBoot) &&
         value <= static_cast<uint16_t>(EventCode::kHistoryCleared);
}

bool valid_event_severity(uint8_t value) {
  return value <= static_cast<uint8_t>(EventSeverity::kCritical);
}

bool valid_event_source(uint8_t value) {
  return value <= static_cast<uint8_t>(EventSource::kGtb);
}

}  // namespace

uint16_t crc16_ccitt(const uint8_t* data, size_t length) {
  uint16_t crc = 0xffffU;
  for (size_t index = 0U; index < length; ++index) {
    crc ^= static_cast<uint16_t>(data[index]) << 8U;
    for (uint8_t bit = 0U; bit < 8U; ++bit) {
      crc = (crc & 0x8000U) != 0U ? static_cast<uint16_t>((crc << 1U) ^ 0x1021U)
                                  : static_cast<uint16_t>(crc << 1U);
    }
  }
  return crc;
}

bool encode_history_record(const HistorySample& sample, HistoryRecord& output) {
  if (sample.output_pct < 0.0F || sample.output_pct > 100.0F) return false;
  const auto output_tenths = static_cast<uint16_t>(std::lround(sample.output_pct * 10.0F));
  if (output_tenths > 1000U) return false;

  output.bytes.fill(0U);
  uint8_t* data = output.bytes.data();
  write_u32_le(data, 0U, sample.sequence);
  write_u32_le(data, 4U, sample.epoch_seconds);
  write_u32_le(data, 8U, sample.uptime_seconds);
  write_u16_le(data, 12U, sample.boot_id);
  write_u16_le(data, 14U, sample.co2_ppm);
  write_u16_le(data, 16U, output_tenths);
  data[18U] = static_cast<uint8_t>(sample.state);
  data[19U] = static_cast<uint8_t>(sample.fault);
  write_u16_le(data, 20U, sample.flags);
  write_u16_le(data, 22U, crc16_ccitt(data, 22U));
  return valid_state(data[18U]) && valid_fault(data[19U]);
}

HistoryDecodeResult decode_history_record(const uint8_t* data, size_t length) {
  HistoryDecodeResult result{false, LogDecodeReason::kLength, {}};
  if (data == nullptr || length != kHistoryRecordSize) return result;

  const uint16_t expected_crc = read_u16_le(data, 22U);
  if (expected_crc != crc16_ccitt(data, 22U)) {
    result.reason = LogDecodeReason::kCrc;
    return result;
  }
  if (!valid_state(data[18U]) || !valid_fault(data[19U])) {
    result.reason = LogDecodeReason::kEnum;
    return result;
  }

  result.ok = true;
  result.reason = LogDecodeReason::kOk;
  result.sample.sequence = read_u32_le(data, 0U);
  result.sample.epoch_seconds = read_u32_le(data, 4U);
  result.sample.uptime_seconds = read_u32_le(data, 8U);
  result.sample.boot_id = read_u16_le(data, 12U);
  result.sample.co2_ppm = read_u16_le(data, 14U);
  result.sample.output_pct = static_cast<float>(read_u16_le(data, 16U)) / 10.0F;
  result.sample.state = static_cast<HistoryState>(data[18U]);
  result.sample.fault = static_cast<HistoryFault>(data[19U]);
  result.sample.flags = read_u16_le(data, 20U);
  return result;
}

bool encode_event_record(const EventEntry& event, EventRecord& output) {
  output.bytes.fill(0U);
  uint8_t* data = output.bytes.data();
  write_u32_le(data, 0U, event.sequence);
  write_u32_le(data, 4U, event.epoch_seconds);
  write_u32_le(data, 8U, event.uptime_seconds);
  write_u16_le(data, 12U, event.boot_id);
  write_u16_le(data, 14U, static_cast<uint16_t>(event.code));
  data[16U] = static_cast<uint8_t>(event.severity);
  data[17U] = static_cast<uint8_t>(event.source);
  write_u16_le(data, 18U, event.count);
  write_i32_le(data, 20U, event.detail0);
  write_i32_le(data, 24U, event.detail1);
  write_u16_le(data, 28U, 0U);
  write_u16_le(data, 30U, crc16_ccitt(data, 30U));
  return valid_event_code(read_u16_le(data, 14U)) &&
         valid_event_severity(data[16U]) &&
         valid_event_source(data[17U]);
}

EventDecodeResult decode_event_record(const uint8_t* data, size_t length) {
  EventDecodeResult result{false, LogDecodeReason::kLength, {}};
  if (data == nullptr || length != kEventRecordSize) return result;

  if (read_u16_le(data, 28U) != 0U) {
    result.reason = LogDecodeReason::kReserved;
    return result;
  }
  const uint16_t expected_crc = read_u16_le(data, 30U);
  if (expected_crc != crc16_ccitt(data, 30U)) {
    result.reason = LogDecodeReason::kCrc;
    return result;
  }
  if (!valid_event_code(read_u16_le(data, 14U)) ||
      !valid_event_severity(data[16U]) ||
      !valid_event_source(data[17U])) {
    result.reason = LogDecodeReason::kEnum;
    return result;
  }

  result.ok = true;
  result.reason = LogDecodeReason::kOk;
  result.event.sequence = read_u32_le(data, 0U);
  result.event.epoch_seconds = read_u32_le(data, 4U);
  result.event.uptime_seconds = read_u32_le(data, 8U);
  result.event.boot_id = read_u16_le(data, 12U);
  result.event.code = static_cast<EventCode>(read_u16_le(data, 14U));
  result.event.severity = static_cast<EventSeverity>(data[16U]);
  result.event.source = static_cast<EventSource>(data[17U]);
  result.event.count = read_u16_le(data, 18U);
  result.event.detail0 = read_i32_le(data, 20U);
  result.event.detail1 = read_i32_le(data, 24U);
  return result;
}

LogStorageEstimate estimate_log_storage(uint16_t sample_period_seconds,
                                        uint16_t retention_days,
                                        uint16_t event_capacity) {
  if (sample_period_seconds == 0U) sample_period_seconds = kHistoryDefaultSamplePeriodSeconds;
  const uint32_t sample_count =
      (static_cast<uint32_t>(retention_days) * 86400UL + sample_period_seconds - 1UL) /
      sample_period_seconds;
  const uint32_t history_bytes = sample_count * kHistoryRecordSize;
  const uint32_t event_bytes = static_cast<uint32_t>(event_capacity) * kEventRecordSize;
  return LogStorageEstimate{sample_count, history_bytes, event_bytes, history_bytes + event_bytes};
}

const char* to_string(HistoryState state) {
  switch (state) {
    case HistoryState::kStartup: return "STARTUP";
    case HistoryState::kAuto: return "AUTO";
    case HistoryState::kTest: return "TEST";
    case HistoryState::kForceOpen: return "FORCE_OPEN";
    case HistoryState::kFault: return "FAULT";
    case HistoryState::kDegradedFlow: return "DEGRADED_FLOW";
    case HistoryState::kUpdate: return "UPDATE";
    case HistoryState::kRecovery: return "RECOVERY";
  }
  return "UNKNOWN";
}

const char* to_string(HistoryFault fault) {
  switch (fault) {
    case HistoryFault::kNone: return "NONE";
    case HistoryFault::kSensorMissing: return "SENSOR_MISSING";
    case HistoryFault::kSensorRange: return "SENSOR_RANGE";
    case HistoryFault::kSensorStale: return "SENSOR_STALE";
    case HistoryFault::kHighCo2: return "HIGH_CO2";
    case HistoryFault::kOutputWriteFailed: return "OUTPUT_WRITE_FAILED";
    case HistoryFault::kConfigCorrupt: return "CONFIG_CORRUPT";
    case HistoryFault::kWifiStartFailed: return "WIFI_START_FAILED";
    case HistoryFault::kHistoryStorageFailed: return "HISTORY_STORAGE_FAILED";
  }
  return "UNKNOWN";
}

const char* to_string(EventCode code) {
  switch (code) {
    case EventCode::kBoot: return "BOOT";
    case EventCode::kConfigChanged: return "CONFIG_CHANGED";
    case EventCode::kConfigRecovery: return "CONFIG_RECOVERY";
    case EventCode::kServiceOpened: return "SERVICE_OPENED";
    case EventCode::kServiceClosed: return "SERVICE_CLOSED";
    case EventCode::kLoginFailed: return "LOGIN_FAILED";
    case EventCode::kTestStarted: return "TEST_STARTED";
    case EventCode::kTestStopped: return "TEST_STOPPED";
    case EventCode::kFaultRaised: return "FAULT_RAISED";
    case EventCode::kFaultCleared: return "FAULT_CLEARED";
    case EventCode::kFirmwareUpdate: return "FIRMWARE_UPDATE";
    case EventCode::kHistoryStorageFailed: return "HISTORY_STORAGE_FAILED";
    case EventCode::kSessionOpened: return "SESSION_OPENED";
    case EventCode::kSessionClosed: return "SESSION_CLOSED";
    case EventCode::kHistoryCleared: return "HISTORY_CLEARED";
  }
  return "UNKNOWN";
}

const char* to_string(EventSeverity severity) {
  switch (severity) {
    case EventSeverity::kInfo: return "INFO";
    case EventSeverity::kWarning: return "WARNING";
    case EventSeverity::kError: return "ERROR";
    case EventSeverity::kCritical: return "CRITICAL";
  }
  return "UNKNOWN";
}

const char* to_string(EventSource source) {
  switch (source) {
    case EventSource::kSystem: return "SYSTEM";
    case EventSource::kPhysical: return "PHYSICAL";
    case EventSource::kInstaller: return "INSTALLER";
    case EventSource::kAdmin: return "ADMIN";
    case EventSource::kGtb: return "GTB";
  }
  return "UNKNOWN";
}

const char* to_string(LogDecodeReason reason) {
  switch (reason) {
    case LogDecodeReason::kOk: return "OK";
    case LogDecodeReason::kLength: return "LENGTH";
    case LogDecodeReason::kCrc: return "CRC";
    case LogDecodeReason::kEnum: return "ENUM";
    case LogDecodeReason::kReserved: return "RESERVED";
  }
  return "UNKNOWN";
}

}  // namespace ventilation
